Userland/Libraries/LibJS/Runtime/DatePrototype.cpp

```diff
@@ -667,43 +671,89 @@ JS_DEFINE_NATIVE_FUNCTION(DatePrototype::to_iso_string)
     return js_string(vm, move(string));
 }
 
-// 21.4.4.38 Date.prototype.toLocaleDateString ( [ reserved1 [ , reserved2 ] ] ), https://tc39.es/ecma262/#sec-date.prototype.tolocaledatestring
+static ThrowCompletionOr<Intl::DateTimeFormat*> construct_date_time_format(GlobalObject& global_object, Value locales, Value options)
+{
+    MarkedValueList arguments { global_object.vm().heap() };
+    arguments.append(locales);
+    arguments.append(options);
+
+    auto* date_time_format = TRY(construct(global_object, *global_object.intl_date_time_format_constructor(), move(arguments)));
+    return static_cast<Intl::DateTimeFormat*>(date_time_format);
+}
+
+// 18.4.2 Date.prototype.toLocaleDateString ( [ locales [ , options ] ] ), https://tc39.es/ecma402/#sup-date.prototype.tolocaledatestring
 JS_DEFINE_NATIVE_FUNCTION(DatePrototype::to_locale_date_string)
 {
+    auto locales = vm.argument(0);
+    auto options = vm.argument(1);
+
+    // 1. Let x be ? thisTimeValue(this value).
     auto* this_object = TRY(typed_this_object(global_object));
+    auto time = this_object->is_invalid() ? js_nan() : this_object->value_of();
```

Comment:
Please don't use `value_of()` (I'm just gonna remove it soon) - it's too close to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf name wise, and it's not clear at all what it actually does: call the prototype method `valueOf()`? Even when overridden by user code? Does that mean it could throw? And so on... (none of these).

`Value(object_ptr)` does the same in a much more understandable way.

Reply:
hmm, so `thisTimeValue` is meant to return the `[[DateValue]]` internal slot, and `value_of` is the only method that does that. I don't think `Value(object_ptr)` works here because we're still left with an object, whereas the `FormatDateTime` AO requires a number.

I could add `Date::date_value` to make the method more explicit?

Reply:
Wow, so it's even worse than I remembered 😄
Yes, those `value_of()`s that wrap an internal member variable in a value should just expose that via an explicit getter instead. Feel free to leave it as is then, I'll take care of it later.

Reply:
FWIW, the one I was thinking about is this:

https://github.com/SerenityOS/serenity/blob/9be409585ca80d2d56112d80d5c4ab2b10aebbd3/Userland/Libraries/LibJS/Runtime/Object.h#L170